### PR TITLE
Tensorpipe based Van implementation for TCP

### PR DESCRIPTION
**Pull request by ZiyueHuang**
It is known that a single TCP connection cannot saturate the bandwidth (e.g., https://github.com/bytedance/byteps/issues/237), and this can be easily verified by `iperf --parallel`. [Tensorpipe](https://github.com/pytorch/tensorpipe) has a channel that multiplexes chunks of data across multiple TCP sockets. 1w1s `./test/test_benchmark 4096000` shows over 50% improvement compared with ZMQVan in terms of Gbps on my testbed. I checked the correctness only by `test_kv_app.cc`. P.S. I only have two nodes for development, each has 16 cpu cores and 0.3 Gbps by a single iperf connection (-_-;). The usage of tensorpipe is mostly adapted from the RPC implementation of pytorch and dgl.

cc @eric-haibin-lin

### CODE REVIEW

**Comment by bobzhuyb** [comment]
This looks promising. Thanks! 
Also cc'ing @ymjiang

**Comment by ymjiang** [comment]
Thank you for the great work! We'll check and give you feedback.

**Comment by ZiyueHuang** [comment]
Thanks for your response.

Here are some info for my building and testing:
```
gcc --version
gcc (GCC) 8.3.1 20190311 (Red Hat 8.3.1-3)
# build tensorpipe
git clone --recursive https://github.com/pytorch/tensorpipe
cd tensorpipe; mkdir build;
cd build; cmake ../;
make
```

```
export DMLC_PS_VAN_TYPE=tp
# For scheduler, DMLC_PS_UV_NTHREADS=1 is enough
# For worker/server, 16 is default in pytorch.rpc
DMLC_PS_UV_NTHREADS=1/16 NUM_KEY_PER_SERVER=xx ./test/test_benchmark 4096000
```

**Comment by ZiyueHuang** [comment]
One potential improvement is to use a `ZPushPull` instead of `ZPush` then `ZPull`. In my understanding, we can save two small messages (`PushResponse` and `PullRequest`) in this way. `ZPushPull` is in [DMLC ps-lite](https://github.com/dmlc/ps-lite/blob/master/include/ps/kv_app.h#L290), but I didn't find this API in BytePS. Is this worth trying in the TCP environment? I didn't see any obvious impediment for using this API, as push is immediately followed by a pull for the same key.

**Comment by ymjiang** [comment]
Thanks for sharing the build script. We keep using Zpush and Zpull APIs since some of our communication ops rely on just a part of them. If the bandwidth can be fully saturated with with tensorpipe, current implementation should be enough.

**Comment by VoVAllen** [comment]
Tensorpipe rocks 🚀

**Comment by ZiyueHuang** [comment]
Ready for review. I have run the tests locally.

I made a few changes: (1) Remove the ZPushPull API. (2) Remove the env variable `DMLC_USE_RECVCTX` and always use separate contexts for send/recv. (3) Improve makefile for TP a little bit. It is safe to build/use ps-lite without TP (the same as before this PR).

Instructions:

To build/install TP (nothing special),
```
cd ~
git clone --recursive https://github.com/pytorch/tensorpipe
cd tensorpipe
mkdir build && mkdir install
cmake -S . -B build
cmake --build build
cmake --install build --prefix ~/tensorpipe/install
```
Note that `~/tensorpipe/install` can be any path suitable to install the library.

To build ps-lite with TP (inside ps-lite directory),
```
USE_TP=1 TP_INSTALL_PATH=~/tensorpipe/install make -j4
```

To use TP as the backend for Van,
```
export DMLC_PS_VAN_TYPE=tp
export DMLC_PS_UV_NTHREADS=xx
```

**Comment by DictXiong** [comment]
Thanks for your contribution! Tests reveal that the tp_van is much faster than the zmq_van. Great job!

I suggest to keep `DMLC_USE_RECVCTX` configurable. Some of my tests find that it would be beneficial not to use separate contexts. When the application does not send and receive data simultaneously and the CPU becomes the bottleneck, using one single context can reduce CPU usage.

Another suggestion is to make the tensorpipe a submodule, and compile it when making ps-lite. The main concern is that if the top application or ps-lite is built with `-fPIC` and a different C++ ABI (`GLIBCXX_USE_CXX11_ABI`), the linking process will fail.

Meanwhile, it is worth mentioning that this branch introduces the env `DMLC_PS_VAN_TYPE` to select the van type. Make sure it is compatible with the old way `DMLC_ENABLE_RDMA`.

**Comment by DictXiong** [comment]
(edit: this doesn't matter if you do not use `DMLC_ENABLE_RDMA` to set van)
I found a critical problem, see: https://github.com/ZiyueHuang/ps-lite/blob/tp/include/ps/kv_app.h#L109
- `is_worker_zpull_` should be `false` when using tp_van
- should support `DMLC_PS_VAN_TYPE`

**Comment by ZiyueHuang** [comment]
Thanks for the review.

> keep DMLC_USE_RECVCTX configurable

Done. This env variable comes back.

> make the tensorpipe a submodule, and compile it when making ps-lite

I think we'd better keep TP as a third party dependency outside ps-lite (similar to UCX), so that users can choose whether to build TP themselves. Besides, TP uses CMake and I found it is not easy to integrate into ps-lite's Makefile. I tried to add TP into deps.mk, but it appears to encounter some wired issues when `make -j` (but make by only one thread seems to work fine). I added a comment in Makefile to remind users about the compile options.

> DMLC_ENABLE_RDMA

As discussed offline, this env variable should be unset when using TCP.

File: tests/test.mk [hunk 1]
```diff
@@ -1,9 +1,14 @@
 TEST_SRC = $(wildcard tests/test_*.cc)
 TEST = $(patsubst tests/test_%.cc, tests/test_%, $(TEST_SRC))
 
+STATIC_LIBS = build/libps.a
+ifeq ($(USE_TP), 1)
+STATIC_LIBS += $(TP_INSTALL_PATH)/lib64/libtensorpipe.a $(TP_INSTALL_PATH)/lib64/libtensorpipe_uv.a
```
Comment (ymjiang) [inline tests/test.mk:6]: Would you also add tensorpipe as a submodule? thanks

Comment (ymjiang) [reply tests/test.mk:6]: Also better to automatically build libtensorpipe.a in pslite makefile.

Comment (ZiyueHuang) [reply tests/test.mk:6]: @ymjiang  I tried to add TP into deps.mk, but it appears to encounter some wired issues when make -j (but make by only one thread seems to work fine). Specifically, the build process terminates after successfully build TP (ps-lite is not being build). So I suggest to build TP as a third party dependency outside ps-lite (similar to UCX). Also, in this way, it is safe to build/use ps-lite without TP (the same as before this PR).

Add \$ (TP) after \$ (ZMQ) in Makefile, and  in deps.mk:
```
TP = ${DEPS_PATH}/include/tensorpipe/tensorpipe.h
${TP}:
    git clone TP and build, as shown in this comment https://github.com/bytedance/ps-lite/pull/111#issuecomment-1380879514.
```

Comment (ymjiang) [reply tests/test.mk:6]: OK, then let's consider fixing this in later MRs. Thank you for you contribution!

**Pull request merged**